### PR TITLE
[ADVAPP-185] & [ADVAPP-186]: Enforce license relationship for Retention CRM with respect to students / prospects

### PR DESCRIPTION
**Pull request by danharrin**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/ADVAPP-185
- https://canyongbs.atlassian.net/browse/ADVAPP-186

### Technical Description

License checking is mainly done in policies and scopes, but there are some exceptions when business logic comes into play.

**The test suite has not been adapted for this changes yet. The tests have not been updated. However, since this is a large PR, it might be good to start reviewing, and then just don't merge until those are passing later today.**

### Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Content or styling update (Changes which don't affect functionality)
- [ ] DevOps
- [ ] Documentation

### Any deployment steps required?

A deployment step could be a command that needs to be executed or an ENV key that needs to be added, for example.

- [x] Yes, please specify
- [ ] No

All users must have licenses. This has been added to the super admin seeder.

_______________________________________________

## Before contributing and submitting this PR, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/canyongbs/advisingapp/blob/main/README.md#contributing).
* [x] Title the PR with the ticket/issue number and a short description of the changes made. Or if no ticket/issue exists, title the PR with a short description of the changes made
* [x] Linked a relevant ticket or issue or describe the issue/feature which this PR resolves/implements.
* [x] Resolved all conflicts, if any.
* [x] Rebased your branch PR on top of the latest upstream `develop` branch.

### CODE REVIEW

File: app-modules/engagement/src/Policies/EngagementPolicy.php [hunk 1]
```diff
@@ -38,10 +38,21 @@
...
+use AdvisingApp\Prospect\Models\Prospect;
 use AdvisingApp\Engagement\Models\Engagement;
+use AdvisingApp\StudentDataModel\Models\Student;
 
 class EngagementPolicy
 {
+    public function before(Authenticatable $authenticatable): ?Response
+    {
+        if (! $authenticatable->hasAnyLicense([Student::getLicenseType(), Prospect::getLicenseType()])) {
+            return Response::deny('You do not have permission to view students or prospects.');
```
Comment (Orrison) [inline app-modules/engagement/src/Policies/EngagementPolicy.php:50]: Same thought to make this generate as above

File: app-modules/notification/src/Policies/SubscriptionPolicy.php [hunk 2]
```diff
@@ -52,6 +52,10 @@ public function viewAny(Authenticatable $authenticatable): Response
 
     public function view(Authenticatable $authenticatable, Subscription $subscription): Response
     {
+        if (! $authenticatable->hasLicense($subscription->subscribable?->getLicenseType())) {
+            return Response::deny('You do not have permission to view this alert.');
```
Comment (Orrison) [inline app-modules/notification/src/Policies/SubscriptionPolicy.php:56]:
```suggestion
            return Response::deny('You do not have permission to view this subscription.');
```

File: app-modules/student-data-model/src/Models/Concerns/BelongsToEducatable.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+use Exception;
+use App\Models\Authenticatable;
+use AdvisingApp\Prospect\Models\Prospect;
+use Illuminate\Database\Eloquent\Builder;
+use AdvisingApp\StudentDataModel\Models\Student;
+use Illuminate\Database\Eloquent\Relations\MorphTo;
+
+trait BelongsToEducatable
+{
+    public function scopeLicensedToEducatable(Builder $query, string $relationship): Builder
```
Comment (Orrison) [inline app-modules/student-data-model/src/Models/Concerns/BelongsToEducatable.php:48]: What would you think about turning this into an Invokable Scope like I did for `EducatableSearch`? They are IMO much better than scope magic methods.

File: app-modules/task/src/Filament/Concerns/TaskForm.php [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+use Filament\Forms\Set;
+use App\Models\Authenticatable;
+use AdvisingApp\Prospect\Models\Prospect;
+use Illuminate\Database\Eloquent\Builder;
+use AdvisingApp\StudentDataModel\Models\Student;
+use Illuminate\Database\Eloquent\Relations\Relation;
+
+trait TaskForm
+{
+    protected function scopeAssignmentRelationshipBasedOnConcern(): Closure
```
Comment (Orrison) [inline app-modules/task/src/Filament/Concerns/TaskForm.php:51]: Same as above, I think Invokable scopes are better for dev experience and better for static analysis.

File: app/Filament/Fields/EducatableSelect.php [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+        $static = app(static::class, ['name' => $name]);
+        $static->configure();
+
+        return $static;
+    }
+
+    public static function getStudentType(): Type
+    {
+        return MorphToSelect\Type::make(Student::class)
```
Comment (Orrison) [inline app/Filament/Fields/EducatableSelect.php:83]: Let's fully qualify this above

File: app/Filament/Fields/EducatableSelect.php [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            ->titleAttribute(Student::displayNameKey());
+    }
+
+    public static function getProspectType(): Type
+    {
+        return MorphToSelect\Type::make(Prospect::class)
```
Comment (Orrison) [inline app/Filament/Fields/EducatableSelect.php:89]: Let's fully qualify this above

File: app/Models/User.php [hunk 7]
```diff
@@ -374,6 +377,81 @@ public function getFilamentAvatarUrl(): ?string
...
+    {
+        if (blank($type)) {
+            return $query;
+        }
+
+        return $query->whereHas(
+            'licenses',
+            fn (Builder $query) => $query->whereIn('type', Arr::wrap($type)),
+        );
+    }
```
Comment (Orrison) [inline app/Models/User.php:453]: Same as the others, my preference would be for these to be invokable scopes.

**Review by Orrison — CHANGES_REQUESTED** [review]

File: app-modules/caseload-management/src/Filament/Resources/CaseloadResource/Pages/CreateCaseload.php [hunk 8]
```diff
@@ -289,9 +297,27 @@ public function getUploadedFileStream(TemporaryUploadedFile $file)
...
+
+        if ($canAccessStudents && $canAccessProspects) {
+            return CaseloadModel::tryFromCaseOrValue($this->form->getRawState()['model']) ?? throw new Exception('Neither students nor prospects were selected.');
+        }
+
+        return match (true) {
+            $canAccessStudents => CaseloadModel::Student,
+            $canAccessProspects => CaseloadModel::Prospect,
+            default => throw new Exception('User cannot access students or prospects.'),
+        };
```
[the remark this answers is not in the document]
Comment (danharrin) [reply app-modules/caseload-management/src/Filament/Resources/CaseloadResource/Pages/CreateCaseload.php:313]: done

File: app-modules/notification/src/Policies/SubscriptionPolicy.php (hunk 2 above)
Comment (danharrin) [reply app-modules/notification/src/Policies/SubscriptionPolicy.php:56]: done

File: app-modules/student-data-model/src/Models/Concerns/BelongsToEducatable.php (hunk 3 above)
Comment (danharrin) [reply app-modules/student-data-model/src/Models/Concerns/BelongsToEducatable.php:48]: done

File: app-modules/task/src/Filament/Concerns/TaskForm.php (hunk 4 above)
Comment (danharrin) [reply app-modules/task/src/Filament/Concerns/TaskForm.php:51]: this is not a scope, its a function that gets shared between the create and edit forms

File: app/Filament/Fields/EducatableSelect.php (hunk 5 above)
Comment (danharrin) [reply app/Filament/Fields/EducatableSelect.php:83]: done

File: app/Filament/Fields/EducatableSelect.php (hunk 6 above)
Comment (danharrin) [reply app/Filament/Fields/EducatableSelect.php:89]: done

File: app/Models/User.php (hunk 7 above)
Comment (danharrin) [reply app/Models/User.php:453]: done

File: app-modules/alert/src/Policies/AlertPolicy.php [hunk 9]
```diff
@@ -39,9 +39,20 @@
...
 use Illuminate\Auth\Access\Response;
+use AdvisingApp\Prospect\Models\Prospect;
+use AdvisingApp\StudentDataModel\Models\Student;
 
 class AlertPolicy
 {
+    public function before(Authenticatable $authenticatable): ?Response
+    {
+        if (! $authenticatable->hasAnyLicense([Student::getLicenseType(), Prospect::getLicenseType()])) {
+            return Response::deny('You do not have permission to view students or prospects.');
```
[the remark this answers is not in the document]
Comment (danharrin) [reply app-modules/alert/src/Policies/AlertPolicy.php:50]: done

File: app-modules/caseload-management/src/Policies/CaseloadPolicy.php [hunk 10]
```diff
@@ -38,10 +38,21 @@
...
+use AdvisingApp\Prospect\Models\Prospect;
+use AdvisingApp\StudentDataModel\Models\Student;
 use AdvisingApp\CaseloadManagement\Models\Caseload;
 
 class CaseloadPolicy
 {
+    public function before(Authenticatable $authenticatable): ?Response
+    {
+        if (! $authenticatable->hasAnyLicense([Student::getLicenseType(), Prospect::getLicenseType()])) {
+            return Response::deny('You do not have permission to view students or prospects.');
```
[the remark this answers is not in the document]
Comment (danharrin) [reply app-modules/caseload-management/src/Policies/CaseloadPolicy.php:50]: done

File: app-modules/engagement/src/Policies/EngagementPolicy.php (hunk 1 above)
Comment (danharrin) [reply app-modules/engagement/src/Policies/EngagementPolicy.php:50]: done

**Review by Orrison — APPROVED** [review]

**Pull request merged**